File: Interview/File System/file_system.cpp
#ifndef __FILE_SYSTEM_CPP__
#define __FILE_SYSTEM_CPP__
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

class FileSystem {
  private:
    struct Node {
        std::string name;
        std::unordered_map<std::string, Node *> children;

        Node(std::string name) : name(name) {}

        Node() : Node("") {}

        bool add_child_from_path(const std::vector<std::string> &path_list,
                                 int start_index = 0) {
            if (start_index == path_list.size()) {
                return false;
            }
            const std::string &child_name = path_list[start_index];
            bool is_child_added           = false;
            if (!contains_child(child_name)) {
                is_child_added       = true;
                children[child_name] = new Node(child_name);
            }
            bool added_in_lower_layer
                = children[child_name]->add_child_from_path(path_list, start_index + 1);
            return is_child_added || added_in_lower_layer;
        }

        void print_tree(std::string indent = "", std::ostream &sout = std::cout) {
            sout << indent << FileSystem::SEPARTOR << name << std::endl;
            for (const auto &[_, child_node] : children) {
                child_node->print_tree(indent + "|\t", sout);
            }
        }

      private:
        bool contains_child(const std::string &name) {
            std::unordered_map<std::string, Node *>::const_iterator child_iterator
                = children.find(name);
            return child_iterator != children.end();
        }
    };

    static std::string SEPARTOR;
    Node virtual_root;

    static std::vector<std::string> split(const std::string &s, const std::string &del) {
        std::vector<std::string> split_string;
        int start = 0;
        int end   = s.find(del);
        while (end != -1) {
            split_string.emplace_back(s.substr(start, end - start));
            start = end + del.size();
            end   = s.find(del, start);
        }
        split_string.emplace_back(s.substr(start, end - start));
        return split_string;
    }

  public:
    bool add_file(const std::string &file_path) {
        return add_file(split(file_path, SEPARTOR));
    }

    bool add_file(const std::vector<std::string> &file_path) {
        return virtual_root.add_child_from_path(file_path, 0);
    }

    void print_as_tree(std::ostream &sout = std::cout) {
        for (const auto &[_, child_node] : virtual_root.children) {
            child_node->print_tree("", sout);
        }
    }
};

std::string FileSystem::SEPARTOR = "/";

#endif // __FILE_SYSTEM__

File: Interview/File System/file_system.py
class FileSystem:

    SEPERATOR = "/"

    class Node:
        def __init__(self, name):
            self.name = name
            self.children = dict()

        def set_children(self, path_array, start_index=0):
            if start_index == len(path_array):
                return False

            children_name = path_array[start_index]
            is_child_added = False
            if children_name not in self.children:
                is_child_added = True
                self.children[children_name] = FileSystem.Node(children_name)

            child_node = self.children.get(children_name)
            added_in_beloy_layer = child_node.set_children(
                path_array, start_index + 1)

            return is_child_added or added_in_beloy_layer

        def print_tree(self, indent=""):
            print(f'{indent}{FileSystem.SEPERATOR}{self.name}')
            for child_node in self.children.values():
                child_node.print_tree(f'{indent}|\t')

    def __init__(self):
        self.virtual_root = FileSystem.Node("")

    def add_file(self, file_path):
        file_path_array = file_path.split(FileSystem.SEPERATOR)
        if not file_path_array:
            return False
        return self.virtual_root.set_children(file_path_array)

    def print_as_tree(self):
        for root in self.virtual_root.children.values():
            root.print_tree()
